ray-operator/controllers/ray/rayservice_controller_test.go

```diff
@@ -483,6 +483,99 @@ applications:
+
+			// Confirm it didn't switch to a new RayCluster
+			Consistently(
+				getRayClusterNameFunc(ctx, myRayService),
+				time.Second*5, time.Millisecond*500).Should(Equal(initialClusterName), "My current RayCluster name  = %v", myRayService.Status.ActiveServiceStatus.RayClusterName)
+			Eventually(
+				getResourceFunc(ctx, client.ObjectKey{Name: myRayService.Status.ActiveServiceStatus.RayClusterName, Namespace: "default"}, myRayCluster),
+				time.Second*3, time.Millisecond*500).Should(BeNil(), "My myRayCluster  = %v", myRayCluster.Name)
+
+			// Verify that the active RayCluster eventually reflects the changes in WorkerGroupSpecs
+			Eventually(
+				getActiveRayClusterWorkerGroupSpecsFunc(ctx, myRayService),
+				time.Second*15,
+				time.Millisecond*500,
+			).Should(
+				HaveLen(2),
+			)
+		})
+		It("should update the pending RayCluster in place when WorkerGroupSpecs are modified by the user in RayServiceSpec", func() {
+			// Trigger a new RayCluster preparation by updating the RayVersion.
+			oldRayVersion := myRayService.Spec.RayClusterSpec.RayVersion
+			newRayVersion := "2.300.0"
+			Expect(oldRayVersion).ShouldNot(Equal(newRayVersion))
+			err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				Eventually(
+					getResourceFunc(ctx, client.ObjectKey{Name: myRayService.Name, Namespace: "default"}, myRayService),
+					time.Second*3, time.Millisecond*500).Should(BeNil(), "My myRayService  = %v", myRayService.Name)
+				myRayService.Spec.RayClusterSpec.RayVersion = newRayVersion
+				return k8sClient.Update(ctx, myRayService)
+			})
+			Expect(err).NotTo(HaveOccurred(), "failed to update test RayService resource with new RayVersion")
+			Eventually(
+				getPreparingRayClusterNameFunc(ctx, myRayService),
+				time.Second*60, time.Millisecond*500).Should(Not(BeEmpty()), "New pending RayCluster name  = %v", myRayService.Status.PendingServiceStatus.RayClusterName)
+			initialPendingClusterName, _ := getPreparingRayClusterNameFunc(ctx, myRayService)()
+
+			// Add a new worker group to the RayServiceSpec
+			newWorkerGroupSpec := myRayService.Spec.RayClusterSpec.WorkerGroupSpecs[0].DeepCopy()
+			newWorkerGroupSpec.GroupName = "worker-group-3"
+
+			err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				Eventually(
+					getResourceFunc(ctx, client.ObjectKey{Name: myRayService.Name, Namespace: "default"}, myRayService),
+					time.Second*3, time.Millisecond*500).Should(BeNil(), "My myRayService  = %v", myRayService.Name)
+				myRayService.Spec.RayClusterSpec.WorkerGroupSpecs = append(myRayService.Spec.RayClusterSpec.WorkerGroupSpecs, *newWorkerGroupSpec)
+				return k8sClient.Update(ctx, myRayService)
+			})
+			Expect(err).NotTo(HaveOccurred(), "failed to update test RayService resource with new WorkerGroupSpecs")
+
+			// Sanity check: length of myRayService.Spec.RayClusterSpec.WorkerGroupSpecs should be 3
+			Expect(myRayService.Spec.RayClusterSpec.WorkerGroupSpecs).Should(HaveLen(3))
```

Comment:
Maybe we should use `len(oldRayService.Spec.RayClusterSpec.WorkerGroupSpecs) + 1` instead.

Reply:
Fixed 8136580